### PR TITLE
fix(python): correct type dispatch in infer_field for custom classes

### PR DESCRIPTION
**Pull request by SURYAS1306**
## Why?

The current type dispatch logic in `infer_field` incorrectly treats certain user-defined classes as built-in types due to insufficient origin/type checks.  
This leads to incorrect field inference and causes Python serialization tests to fail.

## What does this PR do?

This PR refines the type dispatch logic in `infer_field` to properly distinguish between:

- Built-in primitives and standard library types
- Collection types (`list`, `dict`, `set`)
- User-defined classes

User-defined classes are now correctly routed through `visit_customized`, while primitives and supported types continue to use `visit_other`.

All existing Python tests pass after this fix.

## Related issues

None.

## Does this PR introduce any user-facing change?

No.

- [ ] Does this PR introduce any public API change?
- [ ] Does this PR introduce any binary protocol compatibility change?

## Benchmark

Not applicable - this change does not impact performance.

### CODE REVIEW

**Comment by chaokunyang** [comment]
@SURYAS1306 Could you add tests for cases which type inference goes wrong?

File: python/pyfory/format/tests/test_infer.py [hunk 1]
```diff
@@ -63,6 +63,39 @@ class X:
...
+    class Empty:
+        pass
+
+    class WithInit:
+        def __init__(self):
+            self.x = 1
+            self.y = "hello"
+
+    # Should not raise TypeError and must be treated as STRUCT
+    assert _infer_field("", Empty).type.id == TypeId.STRUCT
```
Comment (chaokunyang) [inline python/pyfory/format/tests/test_infer.py]: This may be wrong, only dataclass should be taken as struct. The current struct seiralizer is DataclassSerializer and it only support dataclass too

**Comment by SURYAS1306** [comment]
Hi @chaokunyang,
Thanks for the clarification. I’ve updated the tests to align with the dataclass-only STRUCT behavior and removed the incorrect unannotated class case.

Please let me know if you’d like additional edge cases to be covered.

File: python/pyfory/type_util.py [hunk 2]
```diff
@@ -249,11 +249,17 @@ def infer_field(field_name, type_, visitor: TypeVisitor, types_path=None):
         else:
             raise TypeError(f"Collection types should be {list, dict} instead of {type_}")
     else:
-        if is_function(origin) or not hasattr(origin, "__annotations__"):
+        if is_function(origin):
```
Comment (chaokunyang) [inline python/pyfory/type_util.py:252]: So I do not understand, what does this fix exactly?

Comment (SURYAS1306) [reply python/pyfory/type_util.py:252]: Previously, types without `__annotations__` (including non-dataclass classes) were being routed through the same branch as functions due to the `or not hasattr(origin, "__annotations__")` condition.

This caused incorrect type dispatch for custom classes.

By removing the `not hasattr(origin, "__annotations__")` check, only actual functions are handled in that branch, and class types now follow the intended dispatch path.

Please let me know if I should clarify this further in the PR description.
